Flat config support & improvements for `d.ts`

Hey guys! First of all, thank you for such a cool plugin!

Since ESLint 9 is soon to be released, are there any plans to add a [Flat config](https://eslint.org/docs/latest/use/configure/configuration-files-new) and, equally importantly, improve the `d.ts` files of the plugin?

```
eslint.config.js:25:3 - error TS2322: Type 'typeof import("node_modules/eslint-plugin-wc/lib/index")' is not assignable to type 'Plugin'.
  Types of property 'configs' are incompatible.
    Type '{ recommended: { plugins: string[]; parserOptions: { sourceType: string; }; rules: { 'wc/no-constructor-attributes': string; 'wc/no-invalid-element-name': string; 'wc/no-self-class': string; }; }; 'best-practice': { ...; }; }' is not assignable to type 'Record<string, FlatConfig | ConfigData<RulesRecord> | FlatConfig[]>'.
      Property 'recommended' is incompatible with index signature.
        Type '{ plugins: string[]; parserOptions: { sourceType: string; }; rules: { 'wc/no-constructor-attributes': string; 'wc/no-invalid-element-name': string; 'wc/no-self-class': string; }; }' is not assignable to type 'FlatConfig | ConfigData<RulesRecord> | FlatConfig[]'.
          Type '{ plugins: string[]; parserOptions: { sourceType: string; }; rules: { 'wc/no-constructor-attributes': string; 'wc/no-invalid-element-name': string; 'wc/no-self-class': string; }; }' is not assignable to type 'ConfigData<RulesRecord>'.
            The types of 'parserOptions.sourceType' are incompatible between these types.
              Type 'string' is not assignable to type '"script" | "module" | undefined'.

25   wc,
     ~~

eslint.config.js:27:2 - error TS2322: Type '{ 'wc/guard-super-call': ["off"]; 'wc/attach-shadow-constructor': string; 'wc/no-child-traversal-in-attributechangedcallback': string; 'wc/no-child-traversal-in-connectedcallback': string; ... 5 more ...; 'wc/require-listener-teardown': string; }' is not assignable to type 'RulesRecord'.
  Property ''wc/attach-shadow-constructor'' is incompatible with index signature.
    Type 'string' is not assignable to type 'RuleEntry<any[]>'.

27  rules: {
    ~~~~~

  ../../node_modules/@types/eslint/index.d.ts:1116:9
    1116         rules?: RulesRecord;
                 ~~~~~
    The expected type comes from property 'rules' which is declared here on type 'FlatConfig'
```

--- a/README.md
+++ b/README.md
@@ -36,6 +36,22 @@ Then extend the recommended eslint config:
 }
 ```
 
+Or if you're using ESLint flat configs, add this to your `eslint.config.js`:
+
+```ts
+import {configs} from 'eslint-plugin-wc';
+
+export default [
+  configs.recommended,
+
+  // or if you want to specify `files`, or other options
+  {
+    ...configs.recommended,
+    files: ['test/**/*.js']
+  }
+];
+```
+
 You should also specify settings that will be shared across all the plugin rules. ([More about eslint shared settings](https://eslint.org/docs/user-guide/configuring/configuration-files#adding-shared-settings))
 
 ```jsonc

--- a/src/configs/best-practice.ts
+++ b/src/configs/best-practice.ts
@@ -1,19 +1,22 @@
-const config = {
-  extends: ['plugin:wc/recommended'],
-  plugins: ['wc'],
+import type {ESLint, Linter} from 'eslint';
+import {configFactory as recommendedFactory} from './recommended.js';
 
-  rules: {
-    'wc/attach-shadow-constructor': 'error',
-    'wc/guard-super-call': 'error',
-    'wc/no-child-traversal-in-attributechangedcallback': 'error',
-    'wc/no-child-traversal-in-connectedcallback': 'error',
-    'wc/no-closed-shadow-root': 'error',
-    'wc/no-constructor-params': 'error',
-    'wc/no-customized-built-in-elements': 'error',
-    'wc/no-invalid-extends': 'error',
-    'wc/no-typos': 'error',
-    'wc/require-listener-teardown': 'error'
-  }
+export const configFactory = (plugin: ESLint.Plugin): Linter.FlatConfig => {
+  const base = recommendedFactory(plugin);
+  return {
+    ...base,
+    rules: {
+      ...base.rules,
+      'wc/attach-shadow-constructor': 'error',
+      'wc/guard-super-call': 'error',
+      'wc/no-child-traversal-in-attributechangedcallback': 'error',
+      'wc/no-child-traversal-in-connectedcallback': 'error',
+      'wc/no-closed-shadow-root': 'error',
+      'wc/no-constructor-params': 'error',
+      'wc/no-customized-built-in-elements': 'error',
+      'wc/no-invalid-extends': 'error',
+      'wc/no-typos': 'error',
+      'wc/require-listener-teardown': 'error'
+    }
+  };
 };
-
-export default config;

--- a/src/configs/legacy-best-practice.ts
+++ b/src/configs/legacy-best-practice.ts
@@ -1,0 +1,19 @@
+import type {ESLint} from 'eslint';
+
+export const config: ESLint.ConfigData = {
+  extends: ['plugin:wc/recommended'],
+  plugins: ['wc'],
+
+  rules: {
+    'wc/attach-shadow-constructor': 'error',
+    'wc/guard-super-call': 'error',
+    'wc/no-child-traversal-in-attributechangedcallback': 'error',
+    'wc/no-child-traversal-in-connectedcallback': 'error',
+    'wc/no-closed-shadow-root': 'error',
+    'wc/no-constructor-params': 'error',
+    'wc/no-customized-built-in-elements': 'error',
+    'wc/no-invalid-extends': 'error',
+    'wc/no-typos': 'error',
+    'wc/require-listener-teardown': 'error'
+  }
+};

--- a/src/configs/legacy-recommended.ts
+++ b/src/configs/legacy-recommended.ts
@@ -1,0 +1,11 @@
+import type {ESLint} from 'eslint';
+
+export const config: ESLint.ConfigData = {
+  plugins: ['wc'],
+
+  rules: {
+    'wc/no-constructor-attributes': 'error',
+    'wc/no-invalid-element-name': 'error',
+    'wc/no-self-class': 'error'
+  }
+};

--- a/src/configs/recommended.ts
+++ b/src/configs/recommended.ts
@@ -1,7 +1,8 @@
-const config = {
-  plugins: ['wc'],
-  parserOptions: {
-    sourceType: 'module'
+import type {ESLint, Linter} from 'eslint';
+
+export const configFactory = (plugin: ESLint.Plugin): Linter.FlatConfig => ({
+  plugins: {
+    wc: plugin
   },
 
   rules: {
@@ -9,6 +10,4 @@ const config = {
     'wc/no-invalid-element-name': 'error',
     'wc/no-self-class': 'error'
   }
-};
-
-export default config;
+});

--- a/src/index.ts
+++ b/src/index.ts
@@ -1,5 +1,5 @@
+import type {ESLint} from 'eslint';
 import attachShadowConstructor from './rules/attach-shadow-constructor';
-import bestPractice from './configs/best-practice';
 import defineTagAfterClassDef from './rules/define-tag-after-class-definition';
 import exposeClassOnGlobal from './rules/expose-class-on-global';
 import filenameMatches from './rules/file-name-matches-element';
@@ -19,7 +19,10 @@ import noInvalidExtends from './rules/no-invalid-extends';
 import noOnPrefix from './rules/no-method-prefixed-with-on';
 import noSelfClass from './rules/no-self-class';
 import noTypos from './rules/no-typos';
-import recommended from './configs/recommended';
+import {configFactory as configRecommended} from './configs/recommended';
+import {configFactory as configBestPractice} from './configs/best-practice';
+import {config as configLegacyRecommended} from './configs/legacy-recommended';
+import {config as configLegacyBestPractice} from './configs/legacy-best-practice';
 import requireListenerTeardown from './rules/require-listener-teardown';
 import tagMatchesClass from './rules/tag-name-matches-class';
 
@@ -48,7 +51,11 @@ export const rules = {
   'tag-name-matches-class': tagMatchesClass
 };
 
+const plugin: ESLint.Plugin = {rules};
+
 export const configs = {
-  recommended,
-  'best-practice': bestPractice
+  recommended: configLegacyRecommended,
+  'best-practice': configLegacyBestPractice,
+  'flat/recommended': configRecommended(plugin),
+  'flat/best-practice': configBestPractice(plugin)
 };

--- a/src/test/configs_test.ts
+++ b/src/test/configs_test.ts
@@ -1,0 +1,22 @@
+import type {ESLint, Linter} from 'eslint';
+import {expect} from 'chai';
+import {configs} from '../index';
+
+type ConfigLike = Linter.FlatConfig | ESLint.ConfigData;
+
+const isFlatConfig = (config: ConfigLike): config is Linter.FlatConfig =>
+  !Array.isArray(config.plugins);
+
+describe('configs', () => {
+  it('should define configs correctly', () => {
+    expect(configs['recommended']).to.be.ok;
+    expect(configs['best-practice']).to.be.ok;
+    expect(configs['flat/recommended']).to.be.ok;
+    expect(configs['flat/best-practice']).to.be.ok;
+
+    expect(isFlatConfig(configs['flat/recommended'])).to.equal(true);
+    expect(isFlatConfig(configs['flat/best-practice'])).to.equal(true);
+    expect(isFlatConfig(configs['recommended'])).to.equal(false);
+    expect(isFlatConfig(configs['best-practice'])).to.equal(false);
+  });
+});

